src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/HeapHprofBinWriter.java

```diff
@@ -1274,4 +1301,224 @@ private int getSizeForFields(List<Field> fields) {
+                }
+                return;
+            }
+            super.flush();
+        }
+
+        /**
+         * Enters segmented mode, flush buffered data and set flag.
+         */
+        public void enterSegmentMode() throws IOException {
+            if (allowSegmented && !segmentMode && segmentWritten == 0) {
+                super.flush();
+                segmentMode = true;
+                segmentWritten = 0;
+            }
+        }
+
+        /**
+         * Before finish, flush all data in buffer.
+         */
+        public void finish() throws IOException {
+            if (allowSegmented && segmentMode) {
+                flush();
+                assert segmentWritten == 0;
+                segmentMode = false;
+            }
+        }
+
+        /**
+         * Exits segmented mode, flush segmented data.
+         * @param    force    flush data regardless whether the buffer is full
+         */
+        public void exitSegmentMode() throws IOException {
+            if (allowSegmented && segmentMode && shouldFlush()) {
+                flush();
+                assert segmentWritten == 0;
+                segmentMode = false;
+            }
+        }
+
+
+        /**
+         * Check whether the data should be flush based on data saved in
+         * segmentBuffer.
+         * This method is used to control the segments number and the memory usage.
+         * If segment is too small, there will be lots segments in final dump file.
+         * If it is too large, lots of memory is used in RAM.
+         */
+        private boolean shouldFlush() {
+            // return true if data in segmentBuffer has been extended.
+            return segmentWritten > SEGMENT_BUFFER_SIZE;
+        }
+
+        /**
+         * Writes the write segment header into internal buffer.
+         */
+        private void writeSegmentHeader() {
+            assert segmentWritten == 0;
+            segmentBuffer[segmentWritten++] = (byte)HPROF_HEAP_DUMP_SEGMENT;
+            writeInteger(0);
+            // segment size, write dummy length of 0 and we'll fix it later.
+            writeInteger(0);
+        }
+
+        /**
+         * Fills the segmented data size into the header.
+         */
+        private void fillSegmentSize(int size) {
+            segmentBuffer[5] = (byte)(size >>> 24);
+            segmentBuffer[6] = (byte)(size >>> 16);
+            segmentBuffer[7] = (byte)(size >>>  8);
+            segmentBuffer[8] = (byte)(size >>>  0);
+        }
+
+        /**
+         * Writes an {@code int} to the internal segment buffer
+         * {@code written} is incremented by {@code 4}.
+         */
+        private final void writeInteger(int v) {
+            segmentBuffer[segmentWritten++] = (byte)(v >>> 24);
+            segmentBuffer[segmentWritten++] = (byte)(v >>> 16);
+            segmentBuffer[segmentWritten++] = (byte)(v >>>  8);
+            segmentBuffer[segmentWritten++] = (byte)(v >>>  0);
+        }
```

Comment:
I'm not familiar to hprof format, but I guess we need to write them with big endian.
Can we use `ByteBuffer::putInt` and `ByteBuffer::order` at here? I guess it makes ease to maintenance.

Reply:
Nice Catch! I will make the change, Thanks @YaSuenag !

Reply:
Hi @YaSuenag, 
I found there is also similar code in the original implement, in function fillInHeapRecordLength(). 
` fos.write((dumpLen >>> 24) & 0xFF);`
`  fos.write((dumpLen >>> 16) & 0xFF);`
` fos.write((dumpLen >>> 8) & 0xFF);`       
`  fos.write((dumpLen >>> 0) & 0xFF);`

Do you think it is ok to also change these codes? If my understanding is correct, the fileOutputStream does not have the process of big/little endian. 

Thanks,
Lin

Reply:
> Do you think it is ok to also change these codes? If my understanding is correct, the fileOutputStream does not have the process of big/little endian.

Yes, I agree with your latest change. It looks good to me.